std/stdio.d

```diff
@@ -1074,21 +1134,25 @@ Throws: `ErrnoException` if the file is not opened or if the call to `fwrite` fa
         import std.exception : errnoEnforce;
 
         version (Windows)
-        {
-            flush(); // before changing translation mode
-            immutable fd = ._fileno(_p.handle);
-            immutable mode = ._setmode(fd, _O_BINARY);
-            scope(exit) ._setmode(fd, mode);
-            version (DIGITAL_MARS_STDIO)
-            {
-                import core.atomic : atomicOp;
+        { 
+            // Change mode iff current mode isn't binary
+            if (!(_p.openModeFlags & _p.OpenMode.b))
+            { 
+                flush(); // before changing translation mode
+                immutable fd = ._fileno(_p.handle);
+                immutable mode = ._setmode(fd, _O_BINARY);
+                scope(exit) ._setmode(fd, mode);
```

Comment:
This now happens before the actual write because the `if` introduces a scope relevant for `scope (exit)`.